akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,773 @@
+
+            val diffFastest = fastestConsumer.offsetTimestamp.toEpochMilli - tracking.offsetTimestamp.toEpochMilli
+            val diffFastestStr =
+              if (diffFastest > 0) s"behind fastest [$diffFastest] ms"
+              else if (diffFastest < 0) s"ahead of fastest [$diffFastest] ms" // not possible
+              else "same as fastest"
+            val diffSlowest = slowestConsumer.offsetTimestamp.toEpochMilli - tracking.offsetTimestamp.toEpochMilli
+            val diffSlowestStr =
+              if (diffSlowest > 0) s"behind slowest [$diffSlowest] ms" // not possible
+              else if (diffSlowest < 0) s"ahead of slowest [${-diffSlowest}] ms"
+              else "same as slowest"
+            val consumerBehind = elementsBehind(fastestConsumer.history, tracking.history)
+
+            log.debug(
+              s"Firehose entityType [$entityType] sliceRange [$sliceRangeStr] updateConsumerTracking [$consumerId], " +
+              s"behind [$consumerBehind] events from fastest, " +
+              s"$diffFastestStr, $diffSlowestStr, " +
+              s"slowest [${fastestConsumer.consumerId}] is behind [$behind] " +
+              s"of fastest [${slowestConsumer.consumerId}], [$diffSlowestFastestsMillis] ms. " +
+              s"Consumer lag of fastest [$fastestLagMillis] ms.")
+          }
+        }
+      }
+    }
+
+    def detectSlowConsumers(now: Instant): Unit = {
+      val trackingValues = consumerTrackingValues()
+      if (trackingValues.size > 1) {
+        val slowestConsumer = trackingValues.minBy(_.offsetTimestamp)
+        val fastestConsumer = trackingValues.maxBy(_.offsetTimestamp)
+
+        val behind = elementsBehind(fastestConsumer.history, slowestConsumer.history)
+        val fastestLagMillis = now.toEpochMilli - fastestConsumer.offsetTimestamp.toEpochMilli
+
+        if (behind >= settings.broadcastBufferSize / 2 && fastestLagMillis > 5000) { // FIXME config fastestLagMillis
+          logDetectSlowConsumers(trackingValues, slowestConsumer, fastestConsumer, behind, fastestLagMillis)
+
+          val changedConsumerTrackingValues = trackingValues.flatMap { tracking =>
+            val consumerBehind = elementsBehind(fastestConsumer.history, tracking.history)
+            if (consumerBehind >= settings.broadcastBufferSize / 2) {
+              if (tracking.slowConsumerCandidate.isDefined)
+                None // keep original
+              else
+                Some(tracking.copy(slowConsumerCandidate = Some(now)))
+            } else if (tracking.slowConsumerCandidate.isDefined) {
+              Some(tracking.copy(slowConsumerCandidate = None)) // not slow any more
+            } else {
+              None
+            }
+          }
+
+          changedConsumerTrackingValues.foreach { tracking =>
+            consumerTracking.merge(
+              tracking.consumerId,
+              tracking,
+              (existing, _) => existing.copy(slowConsumerCandidate = tracking.slowConsumerCandidate))
```

Comment:
If it is missing here, doesn't that mean it was stopped/removed, should it be re-introduced in that scenario (memory leak)?

Reply:
you are right, changed in https://github.com/akka/akka/pull/31957/commits/3bb9a9276748079e512d8892028726e59d344992